### PR TITLE
Add a macro returning enum items count

### PR DESCRIPTION
**Pull request by bobeff**
Add a macro `enumLen` which is used to determine the number of items in an enumeration type to the `typetraits.nim` module. Also, add unit tests for it in the `ttypetraits.nim` module.

### CODE REVIEW

**Comment by bobeff** [comment]
I also added a documentation comment and changelog entry which unfortunately I always forget.

**Comment by cooldome** [comment]
Sorry, What is wrong with calling len()? AFAIK, it works for enum typedesc

**Comment by timotheecour** [comment]
> Sorry, What is wrong with calling len()? AFAIK, it works for enum typedesc

```nim
type Foo = enum k1, k2
const s = len(Foo) #  Error: type mismatch: got <type Foo>
```

> elemenetsCount

inconsistent, we have: `len`, `varargsLen`, `tupleLen`
I suggest re-using `len` which has 0 ambiguity (tupleLen vs len was more debatable than for enum). If @araq has a good reason this causes ambiguities, then `enumLen`

File: lib/pure/typetraits.nim
Comment (timotheecour) [inline lib/pure/typetraits.nim]: runnableExamples + also make sure (in tests) that it works and keeps working in presence of various doc comments inside enum definition

Comment (timotheecour) [inline lib/pure/typetraits.nim]: `macro len*(T: typedesc[enum]): int =`

File: lib/pure/typetraits.nim
Comment (timotheecour) [inline lib/pure/typetraits.nim]: `newLit`

File: lib/pure/typetraits.nim
Comment (timotheecour) [inline lib/pure/typetraits.nim:122]: this fails:
```nim
when true:
  import macros
  macro len*(T: typedesc[enum]): int =
    ## Returns the number of items in the enumeration `T`.
    newLit(T.getImpl[2].len - 1)

  type Foo* = enum k0 k1
  static:
    doAssert Foo.len == 2
    type Foo2 = Foo
    doAssert Foo2.len == 2
```

the fix is easy, look at the the other procs in this module

Comment (bobeff) [reply lib/pure/typetraits.nim:122]: Fixed.

File: lib/pure/typetraits.nim
Comment (bobeff) [reply lib/pure/typetraits.nim]: Fixed.

File: lib/pure/typetraits.nim
Comment (bobeff) [reply lib/pure/typetraits.nim]: Fixed.

Comment (bobeff) [reply lib/pure/typetraits.nim]: Fixed.

File: changelog.md
Comment (timotheecour) [inline changelog.md]: rename `elementsCount` :)

File: lib/pure/typetraits.nim
Comment (timotheecour) [inline lib/pure/typetraits.nim]: nit: s/enumeration/enum/ since that's the terminology used

File: lib/pure/typetraits.nim
Comment (timotheecour) [inline lib/pure/typetraits.nim]: fails for `type Foo[T] = enum k0 k1 k2`
`echo Foo[float].len`

this seems to work and is simpler:

```nim
  macro len(T: typedesc[enum]): int =
    var impl = getType(T)
    assert impl.kind == nnkBracketExpr
    impl = impl[1]
    assert impl.kind == nnkEnumTy
    newLit(impl.len - 1)
```

File: changelog.md
Comment (bobeff) [reply changelog.md]: Fixed.

File: lib/pure/typetraits.nim
Comment (bobeff) [reply lib/pure/typetraits.nim]: Fixed.

File: lib/pure/typetraits.nim
Comment (bobeff) [reply lib/pure/typetraits.nim]: Yes, it works. I did it this way and I also added an additional check to the test code to ensure that it works for a type alias chain.

**Review by timotheecour — APPROVED** [review]
only question is len vs enumLen, other than this LGTM

**Comment by Araq** [comment]
One problem here is that there isn't one "len" for enums, there is the "span" (high(E) - low(E) + 1) and then there is the "number of elements". Because enums can have holes, these two can be different.

**Comment by timotheecour** [comment]
> there is the "span" (high(E) - low(E) + 1)

the span would have to be `(high(E).int - low(E).int + 1)` and is almost never what you want with enum with holes, and having `len` map to it would be error prone so I can't see a source of confusion; if there is a use case, users can inline `(high(E).int - low(E).int + 1)` in their code.

That said, `enumLen` is bearable if that's what's needed to unblock this PR.

I also have `len` for `object` (CT for non case object, RT for case object) in some branch.

**Comment by Araq** [comment]
Yeah, let's better use ``enumLen``.

**Comment by Clyybber** [comment]
Maybe `card`?

**Comment by timotheecour** [comment]
> Maybe card?

* `card` is used for `set`, but means a RT value, not CT.
* `enumLen` is more consistent with `varargsLen`, `tupleLen`, `len`, and hopefully in future PR `objLen`.

**Comment by bobeff** [comment]
>     * `enumLen` is more consistent with `varargsLen`, `tupleLen`, `len`, and hopefully in future PR `objLen`.

@timotheecour @Araq But if we have the type of the thing before `len` isn't this a problem for the generic code? Maybe if they all are named simply `len` and we rely only on the overloading to determine the type of the entity which length we are querying it will be easier for the writers of generic libraries.

**Comment by Araq** [comment]
That always comes up ... "generic code", what kind of generic code works with both enums and objects and tuples anyway. And even if does come up, so write ``when T is enum: enumLen(...)``.

The problem is that not every piece of code must compile. Been there, done that, got the bugs. If every type supports the same operations you cannot detect bugs.

**Comment by bobeff** [comment]
I renamed it to `enumLen` and I also squashed the commits into one and I appropriately edited the commit message for cleaner history and commit log.

**Review by timotheecour — APPROVED** [review]

**Comment by cooldome** [comment]
Merging conflict in changelog.md

**Comment by bobeff** [comment]
@cooldome The merge conflict has been resolved.

**Comment by timotheecour** [comment]
no objection to merge?

**Comment by bobeff** [comment]
> no objection to merge?

@timotheecour I have no objections, but of course you are not asking me. :smile: But it is a good moment to ask what exactly is the procedure for merging if any? Two approvals? Approval by @Araq? Something else?

**Comment by nc-x** [comment]
@bobeff 
Afaik, there is no defined process for merging.
But mostly, feature implementations / stdlib additions go through araq (unless it has a `merge_when_passes_ci` label), documentation and tooling stuff usually go through either araq or miran, while bugfix can be merged by anyone with access if it is small, looks good enough and has reasonable amount of tests.

Edit: 
Also, some time back, when this question was raised, https://github.com/nim-lang/Nim/blob/devel/doc/codeowners.rst was created, but I don't think anybody has even seen this file in a long time 😆

**Comment by Araq** [comment]
The process is "Two core devs give their approval (one of them can merge it then) or Araq merges it", will update the docs to say that.

**Pull request merged**
